lalrpop/src/lr1/codegen/parse_table.rs

```diff
@@ -459,12 +493,77 @@ impl<'ascent, 'grammar, W: Write> CodeGenerator<'ascent, 'grammar, W, TableDrive
         rust!(self.out, "return r;");
         rust!(self.out, "}}");
         rust!(self.out, "}} else {{");
+
+        // EOF error recovery
         rust!(self.out,
-              "return Err({}lalrpop_util::ParseError::UnrecognizedToken {{",
+              "let {}error = {}lalrpop_util::ParseError::UnrecognizedToken {{",
+              self.prefix,
               self.prefix);
         rust!(self.out, "token: None,");
         rust!(self.out, "expected: vec![],");
-        rust!(self.out, "}});");
+        rust!(self.out, "}};");
+
+        if self.states.iter().any(|state| state.error.is_some()) {
```

Comment:
Nit: we can we, at minimum, extract this out into a separate fn, but -- better yet -- merge this with `error_recovery`? It seems like the only difference is how it tests whether to keep going at a particular state, right? (i.e., consult `EOF_ACTIONS` vs `ACTIONS`?)